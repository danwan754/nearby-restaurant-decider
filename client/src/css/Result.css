body {
    background-color: #fff2e6;
}

.result-container {
    width: 95%;
    margin: 0 auto;
    text-align: center;
}

.result-detail-container {
    margin: auto;
    border-radius: 7px;
    font-family: Arial, Helvetica, sans-serif;
    background-color: #ffce99;
}

.result-detail-container-table {
    display: table-row;
    margin: auto;
}

.side-result-container {
    display: table-cell;
    vertical-align: top;
    border-radius: 7px;
    padding: 15px;
    height: 80vh;
}

.left-side {
    width: 50%;
}

.right-side {
    width: 44%;
}

.contact-hours-photo-container {
    border-radius: 7px;
    margin: 0;
    padding-bottom: 5px;
    background-color: #fff3e6;
}

.contact-hours-container {
    display: flex;
    padding: 20px;
}

#main-img-container {
    width: 100%;
    height: 125px;
    overflow: hidden;
    position: relative;
    border-radius: 7px;
}
#main-img {
    position: absolute;
    top: -9999px;
    bottom: -9999px;
    left: -9999px;
    right: -9999px;
    margin: auto;
    width: 100%;
}

#restaurant-name {
    font-size: 1.6em;
    font-weight: bold;
    margin-top: 20px;
    padding: 15px;
    text-align: left;
    border-radius: 7px;
    background-color: #fff3e6;
}

#restaurant-hours {
    vertical-align: top;
    text-align: left;
}

#restaurant-contact {
    vertical-align: top;
    text-align: left;
    line-height: 1.5em;
    width: 55%;

}

.rating-time-container {
    display: flex;
    justify-content: space-between;
}

.time-description {
    margin-right: 25px;
}

#skipButton {
    margin: 25px;
    background-color: #ff4d4d;
    color: white;
}

.end-result-container {
    font-size: 1.5em;
}