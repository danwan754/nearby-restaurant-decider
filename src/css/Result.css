
.result-container {
    width: 90%;
    margin: 0 auto;
    text-align: center;
    /* border: 1px solid blue; */
}

.result-detail-container {
    display: table-row; /* attempting to match children heights; remove if fail */

    margin: auto;
    border-radius: 7px;

    border: 1px solid red;
}

.side-result-container {
    /* display: inline-block; */
    display: table-cell; /* attempting to match children heights; remove if fail */
    text-align: left;
    overflow: auto;
    vertical-align: top;
    border-radius: 7px;
    margin: 1%;
    height: 100%;
    
    border: 1px solid black;
}

.left-side {
    width: 50%;
}

.right-side {
    width: 44%;
}

.contact-hours-photo-container {
    padding: 2%;
    border-radius: 7px;
    border: solid black 1px;
}

#restaurant-main-img {
    width: 100%;
    height: 100px;
    vertical-align: top;
    border-radius: 7px;
    /* border: solid 1px black; */
}

#restaurant-name {
    font-size: 1.5em;
    margin-top: 2%;
    text-align: left;
    padding: 10px;
    border-radius: 7px;
    border: 1px black solid;
}

#restaurant-hours {
    display: inline-block;
    vertical-align: top;
    text-align: left;
    margin-left: 5%;
    width: 45%;

    /* border: solid black 1px; */

}

#restaurant-contact {
    display: inline-block;
    vertical-align: top;
    text-align: left;
    width: 45%;

    /* border: solid black 1px; */

}


#restaurant-hours p {
    /* border: solid black 1px; */
    /* margin: 2%; */
}

#restaurant-contact p {
    /* margin: 2%; */
}

.reviews-container {
    text-align: left;
    border: blue solid 1px;
    padding: 10px;
    border-radius: 7px;
    overflow: auto;
    height: 80vh;
}

.review {
    margin: 10px;
    padding: 10px;
    border-radius: 7px;
    border: black 1px solid;
}

.rating-time-container {
    display: flex;
    justify-content: space-between;
}

.time-description {
    margin-right: 25px;
}

.star-container {
}

.star {
    width: 10px;
    height: 10px;
    margin: 2px;
}

.sort-reviews {
    display: flex;
    justify-content: space-between;
    margin-bottom: 20px;
}

.gallery {
    margin: 10px;
    margin-top: 20px;
}

.gallery img {
    width: 25%;
    height: 80px;
}

