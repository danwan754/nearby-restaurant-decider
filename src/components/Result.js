import React from 'react';

import InputBar from './InputBar';
import ResultDetail from './ResultDetail';
import EndResult from './EndResult';
import { getPlaceDetails, getNearByEstablishments, fetchPhoto } from '../requests';

import '../css/Result.css';

class Result extends React.Component {

    state = {
        isLoading: false,
        isEnd: false,
        mainPhotoURL: '',
        errorMessage: '',
        placeIDs: [],
        currentPlaceID: '',
        photoIDs: [],
        place: {}
    }

    constructor() {
        super();
        this.getEstablishments = this.getEstablishments.bind(this);
        this.selectRandomPlace = this.selectRandomPlace.bind(this);
        this.deletePlace = this.deletePlace.bind(this);
        this.getPhotoIDs = this.getPhotoIDs.bind(this);
        this.getNewPlace = this.getNewPlace.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleSkip = this.handleSkip.bind(this);
    }

    componentDidMount() {
        var queryObject = this.props.location ? this.props.location.state : {};

        // query data given by home page
        if (!queryObject){
            return;
        }
        this.getEstablishments(queryObject);
    }

    async getEstablishments(queryObject) {
        var queryString = new URLSearchParams(queryObject).toString();
        this.setState({ isLoading: true });
        try {
            const placeIDs = await getNearByEstablishments(queryString);
            await this.getNewPlace(placeIDs, false);
        }        
        catch(error) {
            this.setState({ 
                errorMessage: error.message,
                isLoading: false 
            });
            return;
        }
    }

    // randomly select a place from list of places
    selectRandomPlace(placeIDs) {
        const length = placeIDs.length;
        if (length === 0) {
            return -1;
        }
        const index = Math.floor((Math.random() * length));
        return index;
    }

    // extract photo IDs from place details object
    getPhotoIDs(placeDetails) {
        let photoIDs = [];
        if (placeDetails.hasOwnProperty('photos')) {
            for (var i = 0; i < placeDetails.photos.length; i++) {
                photoIDs.push(placeDetails.photos[i].photo_reference);
            }
        }
        return photoIDs;
    }

    // delete a place from list of places
    deletePlace() {
        const placeIDs = this.state.placeIDs.filter(place => place !== this.state.currentPlaceID);
        return placeIDs;
    }

    // get new place and update states
    async getNewPlace(placeIDs, isEnd) {
        const index = this.selectRandomPlace(placeIDs);
        const placeDetails = index > -1 ? await getPlaceDetails(placeIDs[index]) : {};
        const photoIDs = this.getPhotoIDs(placeDetails);
        const mainPhotoURL = photoIDs.length > 0 ? await fetchPhoto(photoIDs[0]) : '';
        const currentPlaceID = index > -1 ? placeIDs[index] : '';
        this.setState({ 
            placeIDs: placeIDs,
            currentPlaceID: currentPlaceID,
            place: placeDetails,
            photoIDs: photoIDs,
            mainPhotoURL: mainPhotoURL,
            errorMessage: '',
            isLoading: false,
            isEnd: isEnd
        });
    } 

    handleSubmit(query_object) {
        this.getEstablishments(query_object);
    }

    // delete the current place object from results and show the new place
    async handleSkip() {
        this.setState({ isLoading: true });
        try {
            const placeIDs = this.deletePlace();
            const isEnd = placeIDs.length > 0 ? false : true;
            await this.getNewPlace(placeIDs, isEnd);
        }
        catch(error) {
            this.setState({
                errorMessage: error.message,
                isLoading: false
            });
        }
    }

    render() {
        return(
            <div className="result-container">
                <InputBar
                    query_data={this.props.location.state}
                    onSubmit={this.handleSubmit} />
                { this.state.isLoading ? "Loading ..." :
                    this.state.isEnd ?
                        <EndResult />
                        : 
                        this.state.errorMessage ? this.state.errorMessage 
                            :
                            <React.Fragment>
                                <ResultDetail 
                                    placeID={this.state.currentPlaceID}
                                    place={this.state.place}
                                    photoIDs={this.state.photoIDs}
                                    mainPhotoURL={this.state.mainPhotoURL}
                                    onSubmit={this.handleSkip} />
                                { this.state.placeIDs.length > 0 ?
                                    <input 
                                        id='skipButton'
                                        type='button'
                                        value='Skip'
                                        onClick={ this.handleSkip }
                                        >
                                    </input>
                                    : ''
                                }
                            </React.Fragment>
                }
            </div>
        );
    }
}

export default Result;