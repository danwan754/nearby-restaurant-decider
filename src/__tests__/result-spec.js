import React from 'react';
import TestRenderer from 'react-test-renderer';

import Result from '../components/Result';


describe('Result Component.', () => {
    var resultObj = new Result();

    describe('Inital states are empty.', () => {
    
        it('placeIDs array should have length of zero.', () => {
            expect(resultObj.state.placeIDs.length).toBe(0);
        });

        it('currentPlaceID should be empty string.', () => {
            expect(resultObj.state.currentPlaceID).toBe('');
        });

        it('place object should be empty.', () => {
            const size = Object.keys(resultObj.state.place).length;
            expect(size).toBe(0);
        });
    });

    describe('selectRandomPlace(placeIDs)', () => {
        let placeIDs = ['a', 'b', 'c'];
        let tree;
        let resultInstance;

        beforeEach( () => {
            tree = TestRenderer.create(<Result/>);
            resultInstance = tree.getInstance();
        });

        afterEach( () => {
            tree = null;
        });

        it('Should return null.', () => {
            const emptyPlaceIDs = [];
            const index = resultInstance.selectRandomPlace(emptyPlaceIDs);
            const isNull = index === null;
            expect(isNull).toBe(true);
        });

        it('Should return an index between 0-2 inclusive.', () => {
            const index = resultInstance.selectRandomPlace(placeIDs);
            expect(index >= 0 && index <= 2).toBe(true);
        });

        it('Should update state.currentPlaceID', () => {
            const index = resultInstance.selectRandomPlace(placeIDs);
            const placeID = resultInstance.state.currentPlaceID;
            const isUpdated = placeIDs.includes(placeID);
            expect(isUpdated).toBe(true);
        });

        it('renders correctly', () => {
            tree = tree.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });

    describe('deletePlace()', () => {
        let placeIDs = ['a', 'b', 'c'];
        let place = 'b';
        let tree;
        let resultInstance;

        beforeEach( () => {
            tree = TestRenderer.create(<Result/>);
            resultInstance = tree.getInstance();
        });

        afterEach( () => {
            tree = null;
        });

        it('Should return empty array.', () => {
            const emptyPlaceIDs = [];
            const newPlaceIDs = resultInstance.deletePlace();
            const length = newPlaceIDs.length;
            expect(length).toBe(0);
        });

        it('Should return ["a", "c"].', () => {
            resultInstance.state.placeIDs = ['a', 'b', 'c'];
            resultInstance.state.currentPlaceID = 'b';
            const newPlaceIDs = resultInstance.deletePlace();
            let expected = ['a', 'c'];
            const isSame = JSON.stringify(expected) == JSON.stringify(newPlaceIDs) ? true : false;
            expect(isSame).toBe(true);
        });

        it('Returned value should match state.placeIDs.', () => {
            resultInstance.state.placeIDs = ['a', 'b', 'c'];
            resultInstance.state.currentPlaceID = 'b';
            resultInstance.deletePlace();
            let expected = ['a', 'c'];
            const isSame = JSON.stringify(expected) == JSON.stringify(resultInstance.state.placeIDs) ? true : false;
            expect(isSame).toBe(true);
        })
    }); 
})