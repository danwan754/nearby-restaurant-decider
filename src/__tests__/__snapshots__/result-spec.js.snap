// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result Component. selectRandomPlace(placeIDs) renders correctly 1`] = `
<div
  className="result-container"
>
  <div
    className="input-bar"
  >
    <div
      className="sub-input-bar"
    >
      <div
        className="inline"
      >
        <div>
          <select
            id="dropdown-menu-establishment"
            onChange={[Function]}
          >
            <option>
              Restaurant
            </option>
            <option>
              Cafe
            </option>
            <option>
              Bakery
            </option>
            <option>
              Bar
            </option>
          </select>
        </div>
      </div>
      <div
        className="inline"
      >
        <div>
          <input
            id="input-radius"
            max="5000"
            min="500"
            onChange={[Function]}
            step="500"
            type="number"
            value={2000}
          />
        </div>
      </div>
      <div
        className="inline"
      >
        <div>
          <input
            id="inputPostalCode"
            maxLength="6"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className=""
      >
        <div
          className="country-container"
        >
          <img
            alt="No image"
            className="country-img"
            src="canada.png"
          />
        </div>
      </div>
    </div>
    <input
      className="input-submit inline"
      onClick={[Function]}
      type="submit"
      value="Find"
    />
  </div>
</div>
`;
